_posts/2016-11-14-JS-Glossary.md

```diff
@@ -0,0 +1,554 @@
+
+JavaScript now has a built-in a Promise API, this means every library can work to one API when handling any kind of asynchronous task. I'm not sure what ECMA Specification brought it in. This makes it really easy to make consistent code between libraries. However, more importantly, it makes it possible to have async/await.
+
+### Async/Await
+
+Once Promises were in, then a language construct could be created for using them elegantly. They work by declaring the entire function to be an `async` function. An async function is a function which pretends to be synchronous, but behind the scenes is waiting for specific promises to resolve asynchronously.
+
+There are a few rules for an `async` function:
+
+* You cannot use `await` inside a function that has not been declared `async`. 
+* Anything you do return will be implicitly wrapped in a Promise
+* Non-async functions can just handle the promise an `async` function returns  
+
+So, a typical `async` function
+
+```js
+  async getReviewInfo() : Promise<any> {              // this function will do async
+    const details = await this.getPullRequestInfo()   // wait for the promise in getPullRequestInfo to resolve 
+    return await details.json()                       // wait for the promise in json to resolve
+  }                                                   // return the json
+```
+
+You aren't always given a promise to work with as not all APIs support promises and callbacks, wrapping a callback function is pretty simple: 
+
+```js
+  readFile(path: String): Promise<string> {                       // returns a promise with a string
+    return new Promise((resolve: any, reject: any) => {           // create a new promise, with 2 callbacks
+      fs.readFile(path, "utf8", (err: Error, data: string) => {   // do the work
+        if (err) { return reject(err) }                           // did it fail? reject the promise
+        resolve(data)                                             // did it pass? resolve the promise
+      })
+    })
+  }
+```
+
+The `await` part of an `async` function using `await readFile` will now wait on the synchronous execution until the promise has resolved. This makes complicated code look very simple.
+
+# Types
+
+Types can provide an amazing developer experience, as an editor can understand the shape of all the object's inside your project. This can make it possible to build rich refactoring, static analysis or auto-complete experiences without relying on a runtime.
+
+For JavaScript there are two main ways to use types. [Flow](#flow) and [TypeScript](#typescript). Both are amazing choices for building non-trivial applications. IMO, these two projects are what makes JavaScript a real systems language. 
+
+Both take the approach of providing an optional typing system. This means you can choose to add types to existing applications bit by bit. By doing that you can easily add either to an existing project and progressively add types to unstructured data. 
+
```

Comment:
`By doing that you can easily add either to an existing project and progressively add types to unstructured data.` There is no need for `either` in this sentence.